src/TraceEvent/Computers/TraceManagedProcess.cs

```diff
@@ -32,56 +32,82 @@ namespace Microsoft.Diagnostics.Tracing.Analysis
     /// </summary>
     public static class TraceLoadedDotNetRuntimeExtensions
     {
+        internal static readonly string LoadedDotNetRuntimesKey = "Computers/LoadedDotNetRuntimes";
+
         public static void NeedLoadedDotNetRuntimes(this TraceEventDispatcher source)
         {
+            if (source.userData.ContainsKey(LoadedDotNetRuntimesKey))
+            {
+                // The callbacks are already registered for this source
+                return;
+            }
+
             // ensure there are base processes
```

Comment:
The existing semantics must be persisted, they exist for a reason.   Keep in mind that the state of the items attached to source only evolve during *.Process().  So there is no scenario like the case you mention above.  If NeedLoadedDotNetRuntimes is called multiple times before you call process, it resets, if you call NeedLoadedDotNetRuntimes after calling Process it resets.